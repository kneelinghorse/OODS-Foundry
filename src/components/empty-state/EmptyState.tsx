import * as React from 'react';
import type { ReactNode } from 'react';
import {
  getStatusPresentation,
  type StatusDomain,
  type StatusTone,
} from '../statusables/statusRegistry.js';
import { resolveStatusGlyph } from '../statusables/statusGlyph.js';
import { resolveEmptyStateTokens } from './tokens.js';

export type EmptyStateIntent = 'info' | 'success' | 'warning' | 'neutral';

export interface EmptyStateProps extends React.HTMLAttributes<HTMLDivElement> {
  readonly illustration?: ReactNode;
  readonly icon?: ReactNode;
  readonly headline: ReactNode;
  readonly body?: ReactNode;
  readonly primaryAction?: ReactNode;
  readonly secondaryAction?: ReactNode;
  readonly actions?: ReactNode;
  readonly intent?: EmptyStateIntent;
  readonly status?: string;
  readonly domain?: StatusDomain;
  readonly headlineLevel?: HeadlineLevel;
}

type EmptyStateElement = React.ElementRef<'div'>;
type HeadlineLevel = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

const DEFAULT_DOMAIN: StatusDomain = 'subscription';

const INTENT_TO_TONE: Record<EmptyStateIntent, StatusTone> = {
  info: 'info',
  success: 'success',
  warning: 'warning',
  neutral: 'neutral',
};

export const EmptyState = React.forwardRef<EmptyStateElement, EmptyStateProps>(
  (
    {
      illustration,
      icon,
      headline,
      body,
      primaryAction,
      secondaryAction,
      actions,
      intent = 'neutral',
      status,
      domain: domainProp,
      className,
      headlineLevel = 'h2',
      style,
      ...rest
    },
    forwardedRef
  ) => {
    const domain = (domainProp ?? DEFAULT_DOMAIN) as StatusDomain;
    const statusPresentation = React.useMemo(
      () => (status ? getStatusPresentation(domain, status) : undefined),
      [domain, status]
    );
    const tone: StatusTone = statusPresentation?.tone ?? INTENT_TO_TONE[intent] ?? 'neutral';
    const tokens = React.useMemo(() => resolveEmptyStateTokens(tone), [tone]);
    const resolvedIcon =
      icon ??
      (statusPresentation?.iconName ? resolveStatusGlyph(statusPresentation.iconName) : undefined);
    const rootClassName = className ? `empty-state ${className}` : 'empty-state';
    const styleVariables = React.useMemo<Record<string, string>>(
      () => ({
        '--empty-state-padding': tokens.containerPadding,
        '--empty-state-padding-mobile': tokens.containerPaddingMobile,
        '--empty-state-max-width': tokens.containerMaxWidth,
        '--empty-state-gap': tokens.containerGap,
        '--empty-state-content-gap': tokens.contentGap,
        '--empty-state-illustration-gap': tokens.illustrationGap,
        '--empty-state-illustration-max-width': tokens.illustrationMaxWidth,
        '--empty-state-icon-size': tokens.iconSize,
        '--empty-state-icon-gap': tokens.iconGap,
        '--empty-state-icon-border-radius': tokens.iconBorderRadius,
        '--empty-state-icon-font-size': tokens.iconGlyphSize,
        '--empty-state-icon-background': tokens.iconBackground,
        '--empty-state-icon-foreground': tokens.iconForeground,
        '--empty-state-icon-border': tokens.iconBorder,
        '--empty-state-headline-color': tokens.headlineColor,
        '--empty-state-headline-font-size': tokens.headlineFontSize,
        '--empty-state-headline-font-weight': tokens.headlineFontWeight,
        '--empty-state-headline-line-height': tokens.headlineLineHeight,
        '--empty-state-body-color': tokens.bodyColor,
        '--empty-state-body-font-size': tokens.bodyFontSize,
        '--empty-state-body-line-height': tokens.bodyLineHeight,
        '--empty-state-actions-gap': tokens.actionsGap,
      }),
      [tokens]
    );
    const mergedStyle = React.useMemo<React.CSSProperties>(
      () => ({
        ...styleVariables,
        ...(style ?? {}),
      }),
      [style, styleVariables]
    );
    const hasActions = Boolean(primaryAction || secondaryAction || actions);
    const actionNodes: React.ReactNode[] = React.useMemo(() => {
      const nodes: React.ReactNode[] = [];

      if (primaryAction) {
        nodes.push(
          <div key="primary" className="empty-state__action empty-state__action--primary">
            {primaryAction}
          </div>
        );
      }

      if (secondaryAction) {
        nodes.push(
          <div key="secondary" className="empty-state__action empty-state__action--secondary">
            {secondaryAction}
          </div>
        );
      }

      if (actions) {
        nodes.push(
          <div key="custom" className="empty-state__action empty-state__action--custom">
            {actions}
          </div>
        );
      }

      return nodes;
    }, [actions, primaryAction, secondaryAction]);

    const HeadlineTag = headlineLevel;

    return (
      <div
        ref={forwardedRef}
        className={rootClassName}
        data-intent={intent}
        data-tone={tokens.tone}
        data-status={status ?? undefined}
        data-status-domain={status ? domain : undefined}
        style={mergedStyle}
        {...rest}
      >
        {illustration ? (
          <div className="empty-state__illustration" aria-hidden>
            {illustration}
          </div>
        ) : null}

        {resolvedIcon ? (
          <div className="empty-state__icon" aria-hidden data-tone={tokens.tone}>
            {resolvedIcon}
          </div>
        ) : null}

        <div className="empty-state__content">
          <HeadlineTag className="empty-state__headline">{headline}</HeadlineTag>
          {body ? <p className="empty-state__body">{body}</p> : null}
        </div>

        {hasActions ? <div className="empty-state__actions">{actionNodes}</div> : null}
      </div>
    );
  }
);

EmptyState.displayName = 'OODS.EmptyState';
