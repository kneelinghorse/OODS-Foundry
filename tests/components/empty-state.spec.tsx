/* @vitest-environment jsdom */

/**
 * @file EmptyState component tests
 * @module tests/components/empty-state
 */

import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import '@testing-library/jest-dom/vitest';
import { render, screen, within } from '@testing-library/react';
import { axe } from 'vitest-axe';
import { toHaveNoViolations } from 'vitest-axe/matchers';
import { EmptyState } from '../../src/components/empty-state/EmptyState.js';
import type { EmptyStateProps } from '../../src/components/empty-state/EmptyState.js';

expect.extend({ toHaveNoViolations });

let originalCanvasGetContext: typeof HTMLCanvasElement.prototype.getContext;

beforeAll(() => {
  originalCanvasGetContext = HTMLCanvasElement.prototype.getContext;
  Object.defineProperty(HTMLCanvasElement.prototype, 'getContext', {
    configurable: true,
    value: () => null,
  });
});

afterAll(() => {
  Object.defineProperty(HTMLCanvasElement.prototype, 'getContext', {
    configurable: true,
    value: originalCanvasGetContext,
  });
});

const BASE_PROPS: EmptyStateProps = {
  headline: 'Workspace empty',
  body: 'Create your first record to start seeing data here.',
};

function renderEmptyState(props: Partial<EmptyStateProps> = {}) {
  return render(<EmptyState {...BASE_PROPS} {...props} />);
}

describe('EmptyState', () => {
  describe('structure', () => {
    it('renders headline with default h2 level', () => {
      renderEmptyState();
      const heading = screen.getByRole('heading', { level: 2 });
      expect(heading).toHaveTextContent(BASE_PROPS.headline);
    });

    it('honours the headlineLevel prop', () => {
      renderEmptyState({ headlineLevel: 'h3' });
      const heading = screen.getByRole('heading', { level: 3 });
      expect(heading.tagName.toLowerCase()).toBe('h3');
    });

    it('renders body when provided and omits it when undefined', () => {
      const { rerender, container } = renderEmptyState();
      expect(screen.getByText(BASE_PROPS.body as string)).toBeInTheDocument();

      rerender(<EmptyState {...BASE_PROPS} body={undefined} />);
      expect(container.querySelector('.empty-state__body')).not.toBeInTheDocument();
    });
  });

  describe('slots', () => {
    it('renders the illustration slot as aria-hidden', () => {
      const illustration = <svg data-testid="hero" aria-hidden focusable="false" />;
      const { container } = renderEmptyState({ illustration });
      const illustrationContainer = container.querySelector('.empty-state__illustration');

      expect(screen.getByTestId('hero')).toBeInTheDocument();
      expect(illustrationContainer).toHaveAttribute('aria-hidden');
    });

    it('renders provided icons and hides them from assistive tech', () => {
      const icon = <svg data-testid="status-icon" aria-hidden focusable="false" />;
      const { container } = renderEmptyState({ icon });
      const iconContainer = container.querySelector('.empty-state__icon');

      expect(screen.getByTestId('status-icon')).toBeInTheDocument();
      expect(iconContainer).toHaveAttribute('aria-hidden');
    });
  });

  describe('actions', () => {
    it('omits the actions container when no actions are provided', () => {
      const { container } = renderEmptyState();
      expect(container.querySelector('.empty-state__actions')).not.toBeInTheDocument();
    });

    it('renders primary and secondary actions in order', () => {
      const { container } = renderEmptyState({
        primaryAction: <button type="button">Primary</button>,
        secondaryAction: <button type="button">Secondary</button>,
      });

      const actions = container.querySelector('.empty-state__actions');
      expect(actions).toBeInTheDocument();

      const buttons = within(actions as HTMLElement).getAllByRole('button');
      expect(buttons).toHaveLength(2);
      expect(buttons[0]).toHaveTextContent('Primary');
      expect(buttons[1]).toHaveTextContent('Secondary');
    });

    it('appends custom actions content after slot buttons', () => {
      const { container } = renderEmptyState({
        primaryAction: <button type="button">Create</button>,
        actions: <a href="#" data-testid="support-link">Need help?</a>,
      });

      const actions = container.querySelector('.empty-state__actions');
      expect(actions).toBeInTheDocument();

      const actionChildren = Array.from((actions as HTMLElement).children);
      expect(actionChildren[actionChildren.length - 1]).toContainElement(
        screen.getByTestId('support-link')
      );
    });
  });

  describe('tone and tokens', () => {
    it('applies Statusables tokens for success intent', () => {
      const { container } = renderEmptyState({ intent: 'success' });
      const root = container.firstElementChild as HTMLElement;

      expect(root.dataset.intent).toBe('success');
      expect(root.dataset.tone).toBe('success');
      expect(root.style.getPropertyValue('--empty-state-icon-background')).toBe(
        'var(--cmp-status-success-surface)'
      );
      expect(root.style.getPropertyValue('--empty-state-icon-foreground')).toBe(
        'var(--cmp-status-success-text)'
      );
    });

    it('derives tone and icon from Statusables registry when status is provided', () => {
      const { container } = renderEmptyState({ status: 'trialing', domain: 'subscription' });
      const root = container.firstElementChild as HTMLElement;
      const iconContainer = container.querySelector('.empty-state__icon');

      expect(root.dataset.status).toBe('trialing');
      expect(root.dataset.statusDomain).toBe('subscription');
      expect(root.dataset.tone).toBe('accent');
      expect(iconContainer).toHaveTextContent('â˜…');
      expect(root.style.getPropertyValue('--empty-state-icon-background')).toBe(
        'var(--cmp-status-accent-surface)'
      );
      expect(root.style.getPropertyValue('--empty-state-icon-foreground')).toBe(
        'var(--cmp-status-accent-text)'
      );
    });
  });

  describe('accessibility', () => {
    it('passes axe audit for a fully composed state', async () => {
      const illustration = <svg aria-hidden focusable="false" role="presentation" />;
      const { container } = renderEmptyState({
        illustration,
        intent: 'info',
        primaryAction: <button type="button">Primary action</button>,
        secondaryAction: <button type="button">Secondary action</button>,
      });

      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });

  describe('snapshot', () => {
    it('matches snapshot for a configured empty state', () => {
      const illustration = <svg aria-hidden focusable="false" role="presentation" />;
      const { container } = renderEmptyState({
        illustration,
        intent: 'warning',
        headlineLevel: 'h3',
        primaryAction: <button type="button">Retry</button>,
        secondaryAction: <button type="button">Contact support</button>,
      });

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
