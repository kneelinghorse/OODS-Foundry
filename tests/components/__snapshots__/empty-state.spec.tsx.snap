// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmptyState > snapshot > matches snapshot for a configured empty state 1`] = `
<div
  class="empty-state"
  data-intent="warning"
  data-tone="warning"
  style="--empty-state-padding: var(--cmp-spacing-inset-spacious, var(--sys-spacing-inset-xl, 2rem)); --empty-state-padding-mobile: var(--cmp-spacing-inset-default, var(--sys-spacing-inset-md, 1.5rem)); --empty-state-max-width: 36rem; --empty-state-gap: var(--cmp-spacing-stack-default, 1.5rem); --empty-state-content-gap: var(--cmp-spacing-stack-compact, 0.75rem); --empty-state-illustration-gap: var(--cmp-spacing-stack-default, 1.5rem); --empty-state-illustration-max-width: 16rem; --empty-state-icon-size: 3rem; --empty-state-icon-gap: var(--cmp-spacing-stack-compact, 1rem); --empty-state-icon-border-radius: var(--sys-shape-radius-circle, 50%); --empty-state-icon-font-size: 1.5rem; --empty-state-icon-background: var(--cmp-status-warning-surface); --empty-state-icon-foreground: var(--cmp-status-warning-text); --empty-state-icon-border: var(--cmp-status-warning-border); --empty-state-headline-color: var(--cmp-text-body, var(--sys-color-text-primary, #171717)); --empty-state-headline-font-size: var(--sys-font-size-heading-md, 1.25rem); --empty-state-headline-font-weight: var(--sys-font-weight-semibold, 600); --empty-state-headline-line-height: var(--sys-font-lineheight-heading, 1.3); --empty-state-body-color: var(--cmp-text-muted, var(--sys-color-text-secondary, #525252)); --empty-state-body-font-size: var(--sys-font-size-body-md, 0.875rem); --empty-state-body-line-height: var(--sys-font-lineheight-body, 1.5); --empty-state-actions-gap: var(--cmp-spacing-inline-sm, 0.75rem);"
>
  <div
    aria-hidden="true"
    class="empty-state__illustration"
  >
    <svg
      aria-hidden="true"
      focusable="false"
      role="presentation"
    />
  </div>
  <div
    class="empty-state__content"
  >
    <h3
      class="empty-state__headline"
    >
      Workspace empty
    </h3>
    <p
      class="empty-state__body"
    >
      Create your first record to start seeing data here.
    </p>
  </div>
  <div
    class="empty-state__actions"
  >
    <div
      class="empty-state__action empty-state__action--primary"
    >
      <button
        type="button"
      >
        Retry
      </button>
    </div>
    <div
      class="empty-state__action empty-state__action--secondary"
    >
      <button
        type="button"
      >
        Contact support
      </button>
    </div>
  </div>
</div>
`;
